### PR TITLE
Add giveaway blacklist feature

### PR DESCRIPTION
**Pull request by hilmoo**
- [x] testing

### CODE REVIEW

**Comment by hilmoo** [comment]
this pull req extend #136

**Comment by hilmoo** [comment]
possible error that will happen:
- if someone using the slash command perfectly inside gap task loop
- then some member should removed from table in the next task iteration
- but the command will insert the new same member that should removed in point 2
- there will be a duplicate (or an error since user_id is unique)

**Comment by hilmoo** [comment]
waiting for #136 to merge for testing

File: bot/data/db.sql [hunk 1]
```diff
@@ -73,4 +73,17 @@ CREATE TABLE temp_role (
     PRIMARY KEY(id),
     UNIQUE(user_id, role_id)
 );
-CREATE INDEX IF NOT EXISTS temp_role_user_role_idx ON temp_role (user_id, role_id);
\ No newline at end of file
+CREATE INDEX IF NOT EXISTS temp_role_user_role_idx ON temp_role (user_id, role_id);
+
+--- GIVEAWAY BLACKLIST FEATURE ----
+-----------------------------------
+CREATE TABLE black_ga(
```
Comment (Iqrar99) [inline bot/data/db.sql:80]:
```suggestion
CREATE TABLE blacklist_ga(
```

File: bot/cogs/giveaway.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+    @commands.Cog.listener()
+    async def on_connect(self) -> None:
+        self._check_blacklist_ga.start()
+
+    @app_commands.command(name='blacklist', description='Blacklist a user from giveaway')
+    @app_commands.describe(
+        amount='Giveaway prizes. Example: 50000',
+        winner='Winner of the giveaway. Example: 1234567890,0987654321',
+        ghosting='Member who do not claim the giveaway. Example: 1234567890,0987654321',
```
Comment (Iqrar99) [inline bot/cogs/giveaway.py:29]: The example is too long for a parameter description

File: bot/cogs/giveaway.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        interaction: Interaction,
+        amount: int,
+        winner: str,
+        ghosting: Optional[str] = None,
+    ) -> None:
+        await interaction.response.defer()
+        if not interaction.user.guild_permissions.administrator:
+            return await interaction.followup.send('You are not an admin', ephemeral=True)
+
+        winnerss: list[discord.Member] = []
```
Comment (Iqrar99) [inline bot/cogs/giveaway.py:42]:
```suggestion
        winners: list[discord.Member] = []
```
or put it as `winner_list`

File: bot/cogs/giveaway.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        winners = winner.split(',')
+        for win in winners:
+            user = interaction.guild.get_member(int(win))
+            if user is None:
+                return await interaction.followup.send(f'User {win} not found', ephemeral=True)
+            winnerss.append(user)
+
+        ghostss: list[discord.Member] = []
```
Comment (Iqrar99) [inline bot/cogs/giveaway.py:50]:
```suggestion
        ghosts: list[discord.Member] = []
```

or put it as `ghost_list`

File: bot/cogs/giveaway.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+                        f'User {ghost} not found', ephemeral=True
+                    )
+                ghostss.append(user)
+
+        if amount < GiveawayConfig.BIG_GIVEAWAY:
+            winner_day: 15
+            ghosting_day: 7
+        else:
+            winner_day: 30
+            ghosting_day: 15
```
Comment (Iqrar99) [inline bot/cogs/giveaway.py:66]:
```suggestion
        if amount < GiveawayConfig.BIG_GIVEAWAY:
            winner_day = 15
            ghosting_day = 7
        else:
            winner_day = 30
            ghosting_day = 15
```

File: bot/cogs/giveaway.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+
+        end_time_no_streak = datetime.now(timezone.utc) + timedelta(days=winner_day)
+        end_time_streak = datetime.now(timezone.utc) + timedelta(days=winner_day * 2)
+        end_time_ghosting = datetime.now(timezone.utc) + timedelta(days=ghosting_day)
```
Comment (Iqrar99) [inline bot/cogs/giveaway.py:70]: This can be refactored like `now = datetime.now(timezone.utc)`

File: bot/cogs/giveaway.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+        ghostss: list[discord.Member] = []
+        if ghosting:
+            ghosts = ghosting.split(',')
+            for ghost in ghosts:
+                user = interaction.guild.get_member(ghost)
+                if user is None:
+                    return await interaction.followup.send(
+                        f'User {ghost} not found', ephemeral=True
+                    )
```
Comment (Iqrar99) [inline bot/cogs/giveaway.py:58]: Can be refactored like this
```suggestion
            for (ghosts := ghosting.split(',')) in ghosts:
                if (user := interaction.guild.get_member(ghost)):
                    return await interaction.followup.send(
                        f'User {ghost} not found', ephemeral=True
                    )
```

File: bot/cogs/giveaway.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+        end_time_streak = datetime.now(timezone.utc) + timedelta(days=winner_day * 2)
+        end_time_ghosting = datetime.now(timezone.utc) + timedelta(days=ghosting_day)
+        blacklist_role = interaction.guild.get_role(GiveawayConfig.BLACKLIST_ROLE_ID)
+
+        async with self.db_pool.acquire() as conn:
+            streak_user = await conn.fetch(
+                'SELECT user_id FROM black_ga WHERE status_user = 1',
+            )
+
+            for winn in winnerss:
```
Comment (Iqrar99) [inline bot/cogs/giveaway.py:78]: Fix the variable name

File: bot/cogs/giveaway.py [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+                await winn.add_roles(blacklist_role)
+                if win in streak_user:
```
Comment (Iqrar99) [inline bot/cogs/giveaway.py:80]: Doing look up by using `in` clause in `set()` would faster than `list()` or similar iterator object

File: bot/cogs/giveaway.py [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+                        end_time_no_streak,
+                        1,
+                    )
+                logger.info(
+                    f'Added role {blacklist_role.id} to user {winn.id} for {winner_day} days'
+                )
+
+            for ghost in ghostss:
+                await ghost.add_roles(blacklist_role)
+                if ghost in streak_user:
```
Comment (Iqrar99) [inline bot/cogs/giveaway.py:101]: This is the same as above suggestion

File: bot/cogs/giveaway.py [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+                'SELECT user_id FROM black_ga WHERE end_time <= $1 AND has_role = TRUE',
+                datetime.now(timezone.utc),
+            )
+            user_want_delete = await conn.fetch(
+                'SELECT user_id FROM black_ga WHERE has_role = FALSE AND status_user = 0',
+            )
+
+            for user in user_want_remove:
+                user = guild.get_member(int(user['user_id']))
+                if user:
```
Comment (Iqrar99) [inline bot/cogs/giveaway.py:149]:
```suggestion
                if (user := guild.get_member(int(user['user_id']))):
```

File: bot/cogs/giveaway.py [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+        await interaction.response.defer()
+        if not interaction.user.guild_permissions.administrator:
+            return await interaction.followup.send('You are not an admin', ephemeral=True)
+
+        winnerss: list[discord.Member] = []
+        winners = winner.split(',')
+        for win in winners:
+            user = interaction.guild.get_member(int(win))
+            if user is None:
+                return await interaction.followup.send(f'User {win} not found', ephemeral=True)
```
Comment (Iqrar99) [inline bot/cogs/giveaway.py:47]: This can be refactored

**Review by Iqrar99 — CHANGES_REQUESTED** [review]

**Comment by hilmoo** [comment]
The code still has a syntax error. I’ll check it later, but for now, I can't test it on Discord.

File: bot/cogs/giveaway.py (hunk 2 above)
Comment (hilmoo) [reply bot/cogs/giveaway.py]: I don't think it's too long, it helps users understand how to separate them if more than one is entered

**Comment by hilmoo** [comment]
already tested the feature. lgtm

**Review by Iqrar99 — COMMENTED** [review]
Can you provide the demo as well?

File: bot/cogs/giveaway.py [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+
+    @commands.Cog.listener()
+    async def on_connect(self) -> None:
+        self._check_blacklist_ga.start()
+
+    @app_commands.command(name='blacklist', description='Blacklist a user from giveaway')
+    @app_commands.describe(
+        big='Yes if the giveaway is large than Rp50.000',
+        winners='Winner of the giveaway. Example: 1234567890,0987654321',
+        ghosts='Member who do not claim the giveaway. Example: 1234567890,0987654321',
```
Comment (Iqrar99) [inline bot/cogs/giveaway.py:29]: Should state that these two parameters contain discord ID

```suggestion
        winners='Discord IDs of the giveaway winner. Example: 1234567890,0987654321',
        ghosts='Member IDs who do not claim the giveaway. Example: 1234567890,0987654321',
```

**Review by Iqrar99 — CHANGES_REQUESTED** [review]

**Comment by hilmoo** [comment]
> Can you provide the demo as well?

Check the demo on Discord [here](https://discord.com/channels/761644411486339073/774322083319775262/1287060787700957287).

**Review by Iqrar99 — APPROVED** [review]

**Pull request merged**